% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ComparisonFunctions.R
\name{metagenePlotsForComparison}
\alias{metagenePlotsForComparison}
\title{Function to create metage plots for comparison}
\usage{
metagenePlotsForComparison(data, chrommark, tag, savePlotPath = NULL)
}
\arguments{
\item{data}{metagene-object of metagene profile by 
createMetageneProfile() containing 
input and chip profile}

\item{chrommark}{String, chromatin mark to be analysed. 
Has to be one of the following: H3K36me3, H3K4me3, 
H3K79me2, H4K20me1,H2AFZ,H3K27me3, H3K9me3,H3K27ac,
POLR2AphosphoS5, H3K9ac, H3K4me2, H3K9me1, H3K4me1,
H3K79me1, H3K4ac, H3K14ac, H2BK5ac, H2BK120ac, H2BK15ac,
H4K91ac, H4K8ac, H3K18ac, H2BK12ac, H3K56ac, 
H3K23ac, H2AK5ac, H2BK20ac, H4K5ac, H4K12ac, 
H2A.Z, H3K23me2, H2AK9ac, H3T11ph. 
For RNAPOL2 different variants are available: POLR2A 
(for RNAPol2), POLR3G and POLR2AphosphoS2}

\item{tag}{indicating the kind of profile to plot. 
Can be either: "geneBody", "TES" or "TSS".}

\item{savePlotPath}{if set the plot will be saved under 
"savePlotPath". Default=NULL and plot will be forwarded to stdout.}
}
\value{
nothing, creates a figure under 'savePlotPath'
}
\description{
QC-metrics of newly analysed ChIP-seq samples can be compared with 
the reference values of the compendium and enrichment profiles 
can be plotted against pre-computed profiles of published 
datasets. The metagene profiles show the problematic samples signal 
(red line) for the ChIP, for the input and their relative 
enrichment when compared to the compendiumâ€™s mean signal 
(black line) and its 2 x standard error (blue shadow). 
Additionally the function plots the desired QC-metric
as a red dashed line for the sample plotted against the 
reference distribution (density plots) of the 
compendium values stratified by chromatin marks.

metagenePlotsForComparison
}
\examples{
print("Example Code")
## This command is time intensive to run
##To run the example code the user must provide 2 bam files: 
##one for ChIP and one for the input". Here we used ChIP-seq 
##data from ENCODE. Two example files can be downloaded as follows
## get bam file
setwd(tempdir())

\dontrun{
##chip data
system("wget 
https://www.encodeproject.org/files/ENCFF000BLL/@download/ENCFF000BLL.bam")
chipName="ENCFF000BLL"
chip.data=readBamFile(chipName)

##input data
system("wget 
https://www.encodeproject.org/files/ENCFF000BLL/@download/ENCFF000BKA.bam")
inputName="ENCFF000BKA"
input.data=readBamFile(inputName)

## calculate binding characteristics 
chip_binding.characteristics<-spp::get.binding.characteristics(chip.data, 
srange=c(0,500), bin=5,accept.all.tags=TRUE)

## calculate binding characteristics 
input_binding.characteristics<-spp::get.binding.characteristics(input.data, 
srange=c(0,500), bin=5,accept.all.tags=TRUE)

##get chromosome information and order chip and input by it
chrl_final=intersect(names(chip.data$tags),names(input.data$tags))
chip.data$tags=chip.datatags[chrl_final]
chip.data$quality=chip.data$quality[chrl_final]
input.data$tags=input.data$tags[chrl_final]
input.data$quality=input.data$quality[chrl_final]

##remove sigular positions with extremely high tag counts with 
##respect to the neighbourhood
selectedTags=removeLocalTagAnomalies(chip.data, input.data, 
chip_binding.characteristics, input_binding.characteristics)
input.dataSelected=selectedTags$input.dataSelected
chip.dataSelected=selectedTags$chip.dataSelected

##get smoothed tagdensity 
smoothedChip=tagDensity(chip.dataSelected, 
tag.shift=82)
smoothedInput=tagDensity(input.dataSelected, 
tag.shift=82)

Meta_Result=createMetageneProfile(smoothed.densityChip=smoothedChip, 
smoothedInput,tag.shift=82)

metagenePlotsForComparison(data=Meta_Result$geneBody,chrommark="H3K36me3",
tag="geneBody",savePlotPath=getwd())
metagenePlotsForComparison(data=Meta_Result$TSS,chrommark="H3K36me3",
tag="TSS",savePlotPath=getwd())
}
}
