\documentclass[a4paper,10pt]{article}

\usepackage[latin1]{inputenc}
\usepackage[english]{babel}
\usepackage{fontenc}
\usepackage{graphicx}
\usepackage{hyperref}

%\VignetteIndexEntry{ChIC}

\title{The ChIP-seq quality Control package \textbf{ChIC}: 
A short introduction}

\begin{document}
\maketitle

\begin{abstract}
The \textbf{ChI}P-seq quality \textbf{C}ontrol package (ChIC) provides 
functions and data structures to assess the quality of ChIP-seq data. 
The tool computes three different categories of QC metrics: QC-metrics 
designed for narrow-peak profiles and general metrics, 
QC-metrics based on global read distribution 
and QC-metrics from local signal enrichments around annotated genes. 
User-friendly functions allow to perform the analysis with a single command, 
whereas step by step functions are also available 
for more experienced users. The package comes with a large reference 
compendium of precomputed QC-metrics from public ChIP-seq samples. 
Key features of the package are functions for calculating, visualizing and 
creating summary plots of QC-metrics, 
including the comparison of metagene profiles against reference profiles, 
and the comparison of single QC-metrics with the compendium values. 
% the classifier models are also provided, to compute the single value 
%quality score based on our random forest models.
\end{abstract} 

\newpage

\tableofcontents

\newpage
\section{Installation}
Prerequisites of CHIC are the installation of the data package 
\textit{chic.data} via Bioconductor and spp

<<eval=FALSE, echo=TRUE >>=
install.packages("spp",dependencies=TRUE)
@

The tutorial shows example outputs and usage of several functions using an 
example dataset from ENCODE for H3K36me3 (ID: ENCFF000BLL) and its input 
(ID: ENCFF000BKA).
The sample has been flagged by ENCODE with a red audit because of 
low read depth. The bam files can be downloaded from the following links:

\hyperref[ENCFF000BLL]{https://www.encodeproject.org/files/ENCFF000BLL/}

\hyperref[ENCFF000BKA]{https://www.encodeproject.org/files/ENCFF000BKA/}

\section{QC-metrics based on sharp-peak profiles and 
cross-correlation analysis}

\textit{QC\_scoresCC\_PC} is a wrapper function that reads the bam files 
and calculates a number of QC-metrics from cross-correlation analysis 
and from peak-calling. 

<<eval=TRUE, echo=TRUE >>=
require(ChIC)
chipName="ENCFF000BLL"
inputName="ENCFF000BKA"

## mc is number of CPUs for parallelization 
mc=5
## annotationID is the genome assembly used

## savePlotPath: path to the directory where the final plot 
#should be stored
@

\begin{verbatim}
CC_Result=QCscoresCC_PC(chipName=chipName,
                    inputName=inputName, 
                    read_length=36, 
                    mc=5,
                    annotationID="hg19",
                    savePlotPath=getwd())
\end{verbatim}

The wrapper is based on the following functions:
\subsection{Reading BAM files}
The first step reads ChIP-seq data in .bam file-format. The function 
expects exactly two bam files: 
one for the immunoprecipitation (ChIP) and one for the control (Input). 
The "dataPath" should contain the path to the directory in which the bam 
files are stored.

\begin{verbatim}
chip.data=readBamFile(chipName)
input.data=readBamFile(inputName)
\end{verbatim}


\subsection{Calculate QC-metrics from CrossCorrelation analysis}
The next function called by QC\_scoresCC\_PC is 
\textit{calculateCrossCorrelation} to calculate QC-metrics from 
crosscorrelation analysis and other general metrics, e.g. the 
non-redundant fractions of mapped reads. 
An important parameter required by \textit{calculateCrossCorrelation} 
is the binding-characteristics, calculated by 
\textit{spp::get.binding.characteristics} function. 
The binding-characteristics structure provides information about the 
peak separation distance and the cross-correlation profile (for more details 
see \cite{Kharchenko2008}) and is used in the 
\textit{QC\_scoresCC\_PC} wrapper function.

\begin{verbatim}
## calculate cross correlation QC-metrics for ChIP data
crossvalues_Chip<-calculateCrossCorrelation(chip.data,
    chip_binding.characteristics,
    read_length=36,
    savePlotPath=getwd(),
    plotname="ChIP")
\end{verbatim}

savePlotPath sets the path in which the cross-correlation plot (as pdf) 
should be saved. If not provided the plot will be forwarded to default 
DISPLAY. An example of a cross-correlation profile is shown 
in Figure \ref{fig:ccPlot}.

\begin{figure}
\centering
.\includegraphics[scale=0.3]{ChIP_CrossCorrelation.pdf}
\caption{Cross-correlation plot for sample "ENCFF000BLL" }
\label{fig:ccPlot}
\end{figure}

\textit{calculateCrossCorrelation} must be used on both, 
ChIP and Input. The function returns a number of QC-metrics 
\cite{chicpapero}, amongst others the "tag.shift" value which 
represents an input parameter for further steps (i.e. peak-calling 
and metagene calculation). 

<<eval=TRUE, echo=FALSE >>=
load(file = "CC_ScoresChIP.Rdata")
@


<<eval=FALSE, echo=TRUE >>=
str(CC_Result)
final.tag.shift<-CC_Result$tag.shift
@

<<eval=TRUE, echo=FALSE >>=
str(CC_Result_Chip)
final.tag.shift<-CC_Result_Chip$tag.shift
@

\subsection{Calculate QC-metrics from peak calls}
The last set of QC-metrics come from the called peaks using 
\textit{getBindingRegionsScores}. Within the function 
\textit{removeLocalTagAnomalies} is called 
to perform two steps: 
it selects only tags with acceptable alignment quality (disabled in the 
current version) and removes local tag anomalies like regions 
with extremely high tag counts compared to the neighborhood 
(for more details see \cite{Kharchenko2008}). 

\begin{verbatim}
selectedTags=removeLovalTagAnomalies(chip.data,
    input.data,
    chip_binding.characteristics,
    input_binding.characteristics)

chip.dataSelected=selectedTags$chip.dataSelected

## get QC-values from peak calling
bindingScores=getBindingRegionsScores(chip.data,
    input.data, 
    chip.dataSelected,
    input.dataSelected,
    final.tag.shift)
\end{verbatim}

The function \textit{getBindingRegionsScores} returns a number 
of QC-metrics \cite{Chicpaper}.

<<eval=TRUE, echo=FALSE >>=
load(file = "CC_bindingScores.Rdata")
@

<<eval=TRUE, echo=TRUE >>=
str(CC_Result_Chip)
@


\subsection{Profile smoothing}

\begin{verbatim}
smoothed.densityChip=tagDensity(chip.dataSelected,
    final.tag.shift,
    srngl=rngl)
\end{verbatim}
A necessary step is the smoothing (using a Gaussian kernel) of the tag 
profile to obtain the "tag density profile" (for more details see 
\cite{Kharchenko2008}). The tag density profile is needed to 
calculate the next categories of QC-metrics.

\section{QC-metrics on global read distribution and Fingerprint-plot}
This set of values is based on the global read distribution along the 
genome for ChIP and Input data \cite{Diaz2012}. 
The  function \textit{QCscores\_global} reproduces the so-called 
Fingerprint plot (Figure \ref{fig:ChancePlot}) and returns a list 
of 9 QC-metrics that can be sampled from the cumulative distribution of 
the plot. Examples of these metrics are the (a) fraction of bins without 
reads for ChIP and input, (b) the point of maximum distance between 
the ChIP and input (x-coordinate, y-coordinate for ChIP and input, 
the distance calculated as absolute difference between the two y-coordinates, 
the sign of the difference), (c) the fraction of reads in the top 1 
percent of bins with highest coverage for ChIP and input. 

\begin{verbatim}
## get smoothed tagdensities from QCscoresCC_PC()
smoothedDensityInput=CC_Result$TagDensityInput
smoothedDensityChip=CC_Result$TagDensityChip

Ch_Results=QCscores_global(densityChip=smoothedDensityChip,
    densityInput=smoothedDensityInput,
    savePlotPath=getwd())
\end{verbatim}

<<eval=TRUE, echo=FALSE >>=
load(file = "Ch.Rdata")
@

<<eval=TRUE, echo=TRUE >>=
str(Ch_Results)
@

<<eval=TRUE, echo=FALSE >>=
load(file = "FingerPrintPlot.RData")
f_fingerPrintPlot=function(cumChip,cumInput,savePlotPath=NULL)
{    
    plot(cumChip,type="l",col="blue",lwd=2,xlab="Percentage of bins",
        ylab="Percentage of tags",
        main="Fingerprint: global read distribution")

    lines(cumInput,col="red",lwd=2)
    arrowx=cumChip[which.max(abs(cumInput$pj-cumChip$pj)),]$x
    abline(v =arrowx, col = "green",lty=2,lwd=2)
    
    legend("topleft", legend = c("Input","ChIP"), fill = c("red","blue"))
}


<<eval=FALSE, label=fingerPrintPlot, include=FALSE, echo=TRUE >>=
f_fingerPrintPlot(cumSumChip,cumSumInput)
@

\begin{figure}[htbp]
\centering
<<label=fingerPrintPlot, fig=TRUE, echo=FALSE>>=
<<fingerPrintPlot>>
@
\caption{Fingerprint plot of sample ENCFF000BLL and its input ENCFF000BKA.}
\label{fig:ChancePlot}
\end{figure}


\section{Metagene profiles and QC-metrics based on local enrichment}
Metagene plots show the signal enrichment around a region of interest like 
the transcription start site (TSS) or over a predefined set of genes. 
The tag density of the immunoprecipitation is taken over all RefSeg annotated 
human genes, averaged and log2 transformed. 
The same is done for the input. The normalized profile (Figure \ref{fig:TSS}) 
is calculated as the signal enrichment (immunoprecipitation over the input) 
and plotted on the y-axis, whereas the genomic coordinates of the genes like 
the TSS or regions up- and downstream are shown on the x-axis. 
ChIC creates two types of metagene profiles: a non-scaled profile for the 
TSS and transcription end site, and a scaled profile for the entire gene, 
including the gene body like in Figure \ref{fig:GeneBody}. 

\textit{createMetageneProfile} creates the metagene profiles for scaled 
and non-scaled profiles and returns a list with three items: 
"twopoints", "TSS" and "TES". Each item is again a list with the 
metagene-profiles for ChIP and input.

\begin{verbatim}
Meta_Result=createMetageneProfile(smoothedDensityChip,
    smoothedDensityInput,
    tag.shift,
    annotationID="hg19")
\end{verbatim}

The objects can be used to create the final metagene plots and to get 
the respective QC-values for the non-scaled profiles (Figure \ref{fig:TSS}).

<<eval=FALSE, echo=TRUE >>=
## create non-scaled metagene profile around the TSS
TSS_Plot=nonScaledMetageneProfile(Meta_Result$TSS$chip, 
    Meta_Result$TSS$input, 
    tag="TSS")
## create non-scaled metagene profile around the TES
TES_Plot=nonScaledMetageneProfile(Meta_Result$TES$chip,
    Meta_Result$TES$input,
    tag="TES")
@


\begin{figure}
\centering
.\includegraphics[scale=0.5]{Normalized_TSS.pdf}
\caption{Normalized non-scaled metagene profile for the signal enrichment 
around the TSS }
\label{fig:TSS}
\end{figure}

The "twopoint" object can be used to plot the scaled metagene profile 
and to get the respective QC-values:
<<eval=FALSE, echo=TRUE >>=
#create scaled metagene profile
geneBody_Plot=scaledMetageneProfile(Meta_Result$twopoint$chip, 
    Meta_Result$twopoint$input)
@


\begin{figure}
\centering
.\includegraphics[scale=0.3]{ScaledMetaGene_ChIP_Input.pdf}
\caption{Scaled metagene profile for the tag density across annotated 
genes for ChIP (ENCFF000BLL) and Input (ENCFF000BKA) }
\label{fig:GeneBody}
\end{figure}

\section{Quality assessment using the compendium of QC-metrics as reference}
The comprehensive set of QC-metrics, computed over a large set of ChIP-seq 
samples, constitutes in itself a valuable compendium that can be used as a 
reference for comparison to new samples. ChIC provides two functions for 
that: \textit{metagenePlotsForComparison} to compare the metagene plots with 
the compendium and \textit{plotReferenceDistribution} to compare a 
QC-metric with the compendium values.

\subsection{Comparing local enrichment profiles}
The \textit{metagenePlotsForComparison} function is used to compare the 
local enrichment profile to the reference compendium by plotting the 
metagene profile against the expected metagene for the same type of 
chromatin mark \ref{MetaProfileComparison}. The expected metagene is 
provided by the compendium mean (black line) and standard error 
(blue shadow) shown in Figure \ref{fig:MetaProfileComparison}.

<<eval=FALSE, echo=TRUE >>=
metagenePlotsForComparison(chrommark="H3K36me3", 
    Meta_Result$twopoint, 
    Meta_Result$TSS, 
    Meta_Result$TES)
@

\begin{figure}
\centering
.\includegraphics[scale=0.5]{H3K36me3_TWO_norm.pdf}
\caption{The comparison against the compendium (mean and standard error) 
can highlight potential problems in the signal distribution for a specific 
mark, in case of experiment failure.}
\label{fig:MetaProfileComparison}
\end{figure}

\subsection{Comparing QC-metrics with reference values of the compendium}
The plot against the reference compendium of metrics add an extra level of 
information that can be easily used by less experienced users. 
Indeed, the function \textit{plotReferenceDistribution} is helpful to 
visually compare the characteristics of an analysed sample with a 
large number of already published data (Figure \ref{valueDistributionRSC}).

<<eval=FALSE, echo=TRUE >>=
currentRSCforSample=CC_Result_Chip$QCscores_ChIP$CC_RSC
plotReferenceDistribution(chrommark="H3K36me3",
    metricToBePlotted="RSC", 
    currentValue=currentRSCforSample)
@


\begin{figure}
\centering
.\includegraphics[scale=0.5]{PlotValueDistribution.pdf}
\caption{Density plot showing the QC-metric RSC (red dashed line) of 
ENCFF000BLLs against the reference distribution of the compendium 
values stratified by chromatin classes. }
\label{fig:valueDistributionRSC}
\end{figure}



\newpage
\addcontentsline{toc}{section}{References}
\bibliographystyle{unsrt}
\bibliography{ChicLib}


\end{document}

